lib/segment/src/types.rs

```diff
@@ -2756,15 +2798,172 @@ mod tests {
+        let expected = json!({
+            "b": {
+                "e": {
+                    "f": [1],
+                }
+            }
+        });
+        assert_eq!(payload, expected.into());
+    }
+
+    #[test]
+    fn test_payload_selector_exclude() {
+        let payload = json!({
+            "a": 1,
+            "b": {
+                "c": 123,
+                "e": {
+                    "f": [1,2,3],
+                    "g": 7,
+                    "h": "text",
+                    "i": [
+                        {
+                            "j": 1,
+                            "k": 2
+
+                        },
+                        {
+                            "j": 3,
+                            "k": 4
+                        }
+                    ]
+                }
+            }
+        });
+
+        // exclude
+        let selector = PayloadSelector::new_exclude(vec!["a".to_string(), "b.e.f".to_string()]);
+        let payload = selector.process(payload.into());
+
+        // root removal & nested removal
+        let expected = json!({
+            "b": {
+                "c": 123,
+                "e": {
+                    "g": 7,
+                    "h": "text",
+                    "i": [
+                        {
+                            "j": 1,
+                            "k": 2
+
+                        },
+                        {
+                            "j": 3,
+                            "k": 4
+                        }
+                    ]
+                }
+            }
+        });
+        assert_eq!(payload, expected.into());
+    }
+
+    #[test]
+    fn test_payload_selector_duplicate_exclude() {
+        let payload = json!({
+            "a": 1,
+            "b": {
+                "c": 123,
+                "f": [1,2,3,4,5],
+            }
+        });
+
+        // exclude duplicate
+        let selector =
+            PayloadSelector::new_exclude(vec!["b.f[1]".to_string(), "b.f[1]".to_string()]);
+        let payload = selector.process(payload.into());
+
+        // single removal
+        let expected = json!({
+            "a": 1,
+            "b": {
+                "c": 123,
+                "f": [1,3,4,5],
+            }
+        });
+        assert_eq!(payload, expected.into());
+    }
+
+    #[test]
+    fn test_payload_selector_stable_exclude() {
+        let payload = json!({
+            "a": 1,
+            "b": {
+                "c": 123,
+                "f": [1,2,3,4,5],
+            }
+        });
+
+        // exclude duplicate
+        let selector =
+            PayloadSelector::new_exclude(vec!["b.f[1]".to_string(), "b.f[3]".to_string()]);
+        let payload = selector.process(payload.into());
+
+        // stable removal
+        let expected = json!({
+            "a": 1,
+            "b": {
+                "c": 123,
+                "f": [1,3,5],
+            }
+        });
+        assert_eq!(payload, expected.into());
+    }
```

Comment:
Sadly, the new back-to-front sorting you introduced in c39d9a0f5ce00e402c683731f25658dba42fcea1 is not good enough.

It uses naive string sorting. That way we order the following like this:

1. `arr[10]`
2. `arr[2]`

which is problematic. We should order by index integers instead.

Here's an example test that now fails due to this problem:

```rust
    #[test]
    fn test_payload_selector_stable_exclude() {
        let payload = json!({
            "a": 1,
            "b": {
                "c": 123,
                "f": [1,2,3,4,5,6,7,8,9,10,11,12],
            }
        });

        // exclude duplicate
        let selector =
            PayloadSelector::new_exclude(vec!["b.f[2]".to_string(), "b.f[10]".to_string()]);
        let payload = selector.process(payload.into());

        // stable removal
        let expected = json!({
            "a": 1,
            "b": {
                "c": 123,
                "f": [1,2,4,5,6,7,8,9,10,12],
            }
        });
        assert_eq!(payload, expected.into());
    }
```

Reply:
Good catch :+1:

Reply:
TBH, the proposed solution looks worse than this bug.

I propose to leave it as is